### PR TITLE
Request parser

### PR DESCRIPTION
**Pull request by westerbarendregt**
changed minor things

### CODE REVIEW

File: includes/RequestParser.hpp [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+	HTTP1 = 1
+};
+
+enum	e_status {
+	INVALID = -1,
+	SUCCESS = 0,
+	ERROR = 1
+};
+
+static const char *methods[] = {"GET", "HEAD", "POST", "PUT", "DELETE"};
```
Comment (borispeeters) [inline includes/RequestParser.hpp]: Maybe null terminate?

File: includes/RequestParser.hpp [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+		}
+		return ERROR;
+	}
+	
+	int				GetHeaders(int fd, Client& c)
+	{
+		std::string line;
+		
+		for (int i = 0; i < 18; ++i)
+			c.m_headers.push_back("");
```
Comment (borispeeters) [inline includes/RequestParser.hpp]: Maybe we could use a fill constructor to initialize the vector with 18 empty strings, or the assign function.

File: practice_wester/RequestParser.hpp [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+/*                                                        ::::::::            */
+/*   RequestParser.hpp                                  :+:    :+:            */
+/*                                                     +:+                    */
+/*   By: wester <wester@student.codam.nl>             +#+                     */
+/*                                                   +#+                      */
+/*   Created: 2021/02/23 12:17:47 by wester        #+#    #+#                 */
+/*   Updated: 2021/03/01 17:02:21 by wester        ########   odam.nl         */
+/*                                                                            */
+/* ************************************************************************** */
+
```
Comment (borispeeters) [inline practice_wester/RequestParser.hpp]: we should probably delete the practice_wester directory before merging it to main

File: srcs/utils.cpp [hunk 4]
```diff
@@ -116,3 +116,39 @@ uint16_t hostToNetworkShort(uint16_t hostshort) {
...
+	while (ret > 0)
+	{
+		if (extra.find('\n') != std::string::npos)
+		{
+			size_t found = extra.find_first_of('\n');
+			line = extra.substr(0, found);
+			extra = extra.substr(found + 1, extra.size() - found);
+			return 2;
+		}
+		char *buf = (char*) malloc(1024);
```
Comment (borispeeters) [inline srcs/utils.cpp]: do we need malloc here? i think char buf[1024] would avoid mem leaks here.

File: includes/Client.hpp [hunk 5]
```diff
@@ -12,17 +12,23 @@ class	Client
...
 		bool	fullHttpRequest();
+		std::string	m_request;
+		int                     			m_method;
+		char                        		m_path[1024];
+		int                  	    		m_protocol;
+		size_t								m_content_length;
+		std::vector<std::string> 			m_headers;
+		bool								m_if_body;
+		std::string							m_body;
+		bool								m_done;
```
Comment (tomoliv30) [inline includes/Client.hpp:30]: a request struct with all those entries would be preferable I think, and in private. Then Client could have RequestParser as a friend, (would imply changing RequestParser back to a class with static member function)

File: includes/RequestParser.hpp [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+		std::cout << "BODY:" << c.m_body << std::endl << std::endl;
+		if (c.m_done)
+			std::cout << "PARSING IS DONE" << std::endl;
+		else 
+			std::cout << "PARSING IS NOT FINISHED" << std::endl;
+	}
+	
+	void		Parse(Client& c)
+	{
+		int fd = open("parse.txt", O_RDWR | O_CREAT | O_TRUNC, 0644);
```
Comment (tomoliv30) [inline includes/RequestParser.hpp]: Let's avoid opening a file, writing to it, closing it, reopen it and reading from it, and that for every http request when we already have the info in a string.
we have all the 'std::string' member function at our disposal to read from a std::string. 
Even if std::getline is technically forbidden due to the need for 'sstream' (and I should apply that to the ConfigParser), 
you can easily recode it

File: srcs/Client.cpp [hunk 7]
```diff
@@ -3,9 +3,9 @@
 #include "Error.hpp"
 
 Client::Client() 
-	: m_v_server(0), m_socket(-1), m_request(""), m_received(false), m_treated(false), m_sockaddr(), m_addrlen(sizeof(m_sockaddr))
+	: m_request(""), m_method(-1), m_path(""), m_protocol(-1), m_content_length(0), m_headers(0), m_if_body(false), m_body(""), m_done(false), 
+	m_v_server(0), m_socket(-1), m_received(false), m_treated(false), m_sockaddr(), m_addrlen(sizeof(m_sockaddr))
 {
```
Comment (tomoliv30) [inline srcs/Client.cpp:14]: see my comment about a request struct, that could just have its own constructor and be instantiated in one call here

File: practice_wester/RequestParser.hpp (hunk 3 above)
Comment (tomoliv30) [reply practice_wester/RequestParser.hpp]: and move this file to include

File: srcs/utils.cpp [hunk 8]
```diff
@@ -116,3 +116,39 @@ uint16_t hostToNetworkShort(uint16_t hostshort) {
 		return hostshort;
 	return (hostshort & 0x00FF) << 8 | (hostshort & 0xFF00) >> 8;
 }
+
+int		ft_getline(int fd, std::string& line)
+{
```
Comment (tomoliv30) [inline srcs/utils.cpp:121]: See my comment about opening file, here getline() could operate on std::string only. 
And wouldn't it be easier to have getline() just return or update the current line, and other functions to analyze what we have in this line, thus avoiding to return 1 or 2 here, and allowing the function to be used in different context, everytime we need to get the next line in a string.

File: includes/RequestParser.hpp (hunk 1 above)
Comment (westerbarendregt) [reply includes/RequestParser.hpp]: Yeah, probably easier!

File: practice_wester/RequestParser.hpp (hunk 3 above)
Comment (westerbarendregt) [reply practice_wester/RequestParser.hpp]: yes please!

File: srcs/utils.cpp (hunk 4 above)
Comment (westerbarendregt) [reply srcs/utils.cpp]: no we don't I thought I took it out already

File: includes/Client.hpp (hunk 5 above)
Comment (westerbarendregt) [reply includes/Client.hpp:30]: I changed it back to a class indeed, this seems to combine with the struct.

File: includes/RequestParser.hpp (hunk 6 above)
Comment (westerbarendregt) [reply includes/RequestParser.hpp]: indeed!

**Review by borispeeters — APPROVED** [review]
looks good

**Review by tomoliv30 — APPROVED** [review]

**Comment by westerbarendregt** [comment]
I think I merged

**Pull request merged**